# Introduction
In this project, we look to clean and analyse the data set of `used cars` from eBay Kleinanzeigen, a `classifieds` section of the German eBay website. You can take a look at the data set [here](https://data.world/data-society/used-cars-data).

The data dictionary provided as follows:

| Column | Description |
| --- | --- |
|dateCrawled | When this ad was first crawled. All field-values are taken from this date|
|name | Name of the car|
|seller | Whether the seller is private or a dealer|
|offerType | The type of listing|
|price | The price on the ad to sell the car|
|abtest | Whether the listing is included in an A/B test|
|vehicleType | The vehicle Type|
|yearOfRegistration | The year in which the car was first registered|
|gearbox | The transmission type|
|powerPS | The power of the car in PS|
|model | The car model name|
|kilometer | How many kilometers the car has driven|
|monthOfRegistration | The month in which the car was first registered|
|fuelType | What type of fuel the car uses|
|brand | The brand of the car|
|notRepairedDamage | If the car has a damage which is not yet repaired|
|dateCreated | The date on which the eBay listing was created|
|nrOfPictures | The number of pictures in the ad|
|postalCode | The postal code for the location of the vehicle|
|lastSeenOnline | When the crawler saw this ad last online|

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("C:/Users/User/Datasets/autos.csv", encoding = "Windows-1252")

In [2]:
#explore data set with info() and head() methods
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Data cleaning and exploration
We see that there are some null values in certain columns. These columns include `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage`. Also, it is worth noting that the columns are using camelcase instead of snakecase. Let us modify the columns to Python's preferred snakecase.

In [3]:
print(autos.columns)
new_columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns=new_columns
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

After modifying the column headers, let us begin exploring the data set. 

Some things to look out for:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

Commonly used methods include:
- describe() method
- value_counts() method
- isnull() method

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From preliminary observation, it seems that `seller` and `offer_type` column may be redundant as there are only 2 types of values and one of which has only 1 occurrence. We can investigate the other unique value and remove the column if it is not significant. The `nr_of_pictures` column seems to only have '0' as value. Hence, we can also remove this column.

The price column is also worth investigating as there are 1421 occurrences where the price is \\$0. We may want to leave out these rows before further analysis. There are also several columns that we can convert the data type so we can better analyse this data set. These columns include: `date_crawled`, `price`, `odometer` and `last_seen`.

Let us first convert the data type of `price` and `odometer` columns.

In [5]:
#cleaning price column
print("Number of numeric entries in price column before cleaning: \n",autos["price"].str.isnumeric().value_counts(),"\n")
price= autos["price"].str.replace("$","").str.replace(",","")
#check that all prices are numeric
print("Number of numeric entries in price column after cleaning: \n",price.str.isnumeric().value_counts())
#convert data type to numeric
autos["price"]=price.astype(int)
print(autos.info())

Number of numeric entries in price column before cleaning: 
 False    50000
Name: price, dtype: int64 

Number of numeric entries in price column after cleaning: 
 True    50000
Name: price, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   powerPS             50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registrati

In [6]:
#cleaning odometer column
print("Number of numeric entries in odometer column before cleaning: \n",autos["odometer"].str.isnumeric().value_counts(),"\n")
odometer = autos["odometer"].str.replace("km","").str.replace(",","")
print("Number of numeric entries in odometer column after cleaning: \n",odometer.str.isnumeric().value_counts(),"\n")
#convert data type to numeric
autos["odometer"]=odometer.astype(int)
print(autos.info())
#rename column odometer to odometer_km
autos.rename(columns={"odometer":"odometer_km"}, inplace=True)

Number of numeric entries in odometer column before cleaning: 
 False    50000
Name: odometer, dtype: int64 

Number of numeric entries in odometer column after cleaning: 
 True    50000
Name: odometer, dtype: int64 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   powerPS             50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  int32 
 

In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   powerPS             50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int32 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

Now that we have made the changes for `odometer` and `price` columns, let us now investigate further.

## Exploring the price column

In [8]:
#to see number of unique values
print(autos["price"].unique().shape)

#view min/max/mean/median etc
print(autos["price"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [9]:
# Change the format to show floats with 1 decimal place and include a thousand separator
pd.options.display.float_format = '{:,.1f}'.format

In [10]:
print(autos["price"].describe())

count       50,000.0
mean         9,840.0
std        481,104.4
min              0.0
25%          1,100.0
50%          2,950.0
75%          7,200.0
max     99,999,999.0
Name: price, dtype: float64


In [11]:
#look at the values and counts in each column
print(autos["price"].value_counts().sort_index(ascending=True).head(10),"\n")

print(autos["price"].value_counts().sort_index(ascending=False).head(20))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64 

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


It is alarming that the price of a used car is listed as high as \\$9999999 and as low as \$0. 

A quick search on google yields the result that the most expensive [car](https://www.google.com/search?q=most+expensive+car&rlz=1C1CHBF_enSG925SG925&oq=most+expensive+car&aqs=chrome..69i57j0i67j0j0i20i263j0l6.2052j0j7&sourceid=chrome&ie=UTF-8) available in the market prices at $3,900,000.

Since our data set is on used cars, it is unlikely that the car would be in the range of millions.
For the sake of better representation of majority of the cars in the market, it seems more logical to retain data within the price range of \\$1 to \$350,000. This would allow us to maintain the integrity of the data and remove unnecessary outliers that could skew our data. 

In [12]:
autos=autos.loc[autos["price"].between(1,350000),:]

## Exploring the odometer column

In [13]:
print(autos["odometer_km"].describe(),"\n")
print(autos["odometer_km"].unique().shape,"\n")
print(autos["odometer_km"].value_counts().sort_index(ascending=True).head(10))
autos["odometer_km"].value_counts(normalize=True)

count    48,565.0
mean    125,770.1
std      39,788.6
min       5,000.0
25%     125,000.0
50%     150,000.0
75%     150,000.0
max     150,000.0
Name: odometer_km, dtype: float64 

(13,) 

5000      836
10000     253
20000     762
30000     780
40000     815
50000    1012
60000    1155
70000    1217
80000    1415
90000    1734
Name: odometer_km, dtype: int64


150000   0.6
125000   0.1
100000   0.0
90000    0.0
80000    0.0
70000    0.0
60000    0.0
50000    0.0
5000     0.0
40000    0.0
30000    0.0
20000    0.0
10000    0.0
Name: odometer_km, dtype: float64

From the values shown above, `odometer_km` column has reasonable values even though the values from 50 percentile onwards are all 150,000.


## Exploring the registration_year column

In [14]:
autos["registration_year"].describe()

count   48,565.0
mean     2,004.8
std         88.6
min      1,000.0
25%      1,999.0
50%      2,004.0
75%      2,008.0
max      9,999.0
Name: registration_year, dtype: float64

In order for our data to be more relevant, let us remove the rows before 1900 and after 2016. Even though it is not likely that a car would be registered and sold more than a 100 years later, let us just keep to 1900 to filter out other outliers beyond 1900 as the birth year of the modern car is 1886.

In [15]:
50000-autos.loc[autos["registration_year"].between(1900,2016),"registration_year"].value_counts().sum()

3319

On further investigation, we find that there are 3319 rows not in the range of 1900-2016. Let us remove these rows and calculate the distribution of the remaining values.

In [16]:
autos=autos.loc[autos["registration_year"].between(1900,2016),:]

In [17]:
#verify that all rows with registration_year 2017 has been removed
autos.loc[autos["registration_year"]==2017,"registration_year"].value_counts()

Series([], Name: registration_year, dtype: int64)

In [18]:
# Change the format to show floats with 5 decimal place and include a thousand separator
pd.options.display.float_format = '{:,.5f}'.format
autos["registration_year"].value_counts(normalize=True).head(20)

2000   0.06761
2005   0.06289
1999   0.06206
2004   0.05790
2003   0.05782
2006   0.05720
2001   0.05647
2002   0.05326
1998   0.05062
2007   0.04878
2008   0.04745
2009   0.04466
1997   0.04179
2011   0.03477
2010   0.03404
1996   0.02941
2012   0.02806
1995   0.02628
2016   0.02613
2013   0.01720
Name: registration_year, dtype: float64

Majority of the registration_year fall between the 1990s to 2010s. This seems to be fairly consistent as majority of the cars listed are approximately 20 years from the date of registration.

## Exploring single value columns
Let us now explore the columns that have mainly the same value. (seller, offer_type and nr_of_pictures columns)

In [19]:
print(autos["nr_of_pictures"].value_counts(),"\n")
print(autos["seller"].value_counts(),"\n")
print(autos["offer_type"].value_counts(),"\n")
autos.loc[autos["seller"]=="gewerblich"]

0    46681
Name: nr_of_pictures, dtype: int64 

privat        46680
gewerblich        1
Name: seller, dtype: int64 

Angebot    46681
Name: offer_type, dtype: int64 



date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type  price   abtest vehicle_type  registration_year  \
7738  gewerblich    Angebot    100  control        kombi               2000   

      gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
7738  manuell        0  megane       150000                   8    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
7738  renault               NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

To maintain the congruity of data, let us remove the row 7738 before removing the three columns stated above.

In [20]:
#remove row 7738
autos.drop([7738],axis=0,inplace=True)

In [21]:
#drop the three columns
autos.drop(["nr_of_pictures","seller","offer_type"], axis=1,inplace=True)

In [22]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46680 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        46680 non-null  object
 1   name                46680 non-null  object
 2   price               46680 non-null  int32 
 3   abtest              46680 non-null  object
 4   vehicle_type        43976 non-null  object
 5   registration_year   46680 non-null  int64 
 6   gearbox             44570 non-null  object
 7   powerPS             46680 non-null  int64 
 8   model               44487 non-null  object
 9   odometer_km         46680 non-null  int32 
 10  registration_month  46680 non-null  int64 
 11  fuel_type           43362 non-null  object
 12  brand               46680 non-null  object
 13  unrepaired_damage   38374 non-null  object
 14  ad_created          46680 non-null  object
 15  postal_code         46680 non-null  int64 
 16  last_seen           46

Now that we have cleaned up the unnecessary columns and data, let us explore the date columns.
# Date columns
These columns include: 
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Currently, `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Let us look at how we can make use of these columns better for our analysis.

In [23]:
autos[["date_crawled","last_seen","ad_created"]].head(10)

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00
5  2016-03-21 13:47:45  2016-04-06 09:45:21  2016-03-21 00:00:00
6  2016-03-20 17:55:21  2016-03-23 02:48:59  2016-03-20 00:00:00
7  2016-03-16 18:55:19  2016-04-07 03:17:32  2016-03-16 00:00:00
8  2016-03-22 16:51:34  2016-03-26 18:18:10  2016-03-22 00:00:00
9  2016-03-16 13:47:02  2016-04-06 10:46:35  2016-03-16 00:00:00

Let us remove the time stamp so we can analyse the date values.

In [24]:
#reassign the cleaned values back to the columns
autos["date_crawled"]=autos["date_crawled"].str[:10]
autos["last_seen"]=autos["last_seen"].str[:10]
autos["ad_created"]=autos["ad_created"].str[:10]

In [25]:
#further analysis of date_crawled column
autos["date_crawled"].value_counts(normalize=True).sort_index(ascending=True)

2016-03-05   0.02519
2016-03-06   0.01416
2016-03-07   0.03625
2016-03-08   0.03355
2016-03-09   0.03325
2016-03-10   0.03224
2016-03-11   0.03246
2016-03-12   0.03683
2016-03-13   0.01587
2016-03-14   0.03633
2016-03-15   0.03434
2016-03-16   0.02950
2016-03-17   0.03179
2016-03-18   0.01281
2016-03-19   0.03466
2016-03-20   0.03802
2016-03-21   0.03732
2016-03-22   0.03284
2016-03-23   0.03220
2016-03-24   0.02948
2016-03-25   0.03151
2016-03-26   0.03207
2016-03-27   0.03078
2016-03-28   0.03460
2016-03-29   0.03410
2016-03-30   0.03380
2016-03-31   0.03179
2016-04-01   0.03380
2016-04-02   0.03556
2016-04-03   0.03877
2016-04-04   0.03661
2016-04-05   0.01300
2016-04-06   0.00308
2016-04-07   0.00141
Name: date_crawled, dtype: float64

In [26]:
#further analysis of last_seen column
autos["last_seen"].value_counts(normalize=True).sort_index(ascending=True)

2016-03-05   0.00107
2016-03-06   0.00411
2016-03-07   0.00538
2016-03-08   0.00748
2016-03-09   0.00977
2016-03-10   0.01069
2016-03-11   0.01238
2016-03-12   0.02376
2016-03-13   0.00865
2016-03-14   0.01266
2016-03-15   0.01600
2016-03-16   0.01628
2016-03-17   0.02808
2016-03-18   0.00722
2016-03-19   0.01562
2016-03-20   0.02063
2016-03-21   0.02059
2016-03-22   0.02084
2016-03-23   0.01836
2016-03-24   0.01969
2016-03-25   0.01894
2016-03-26   0.01680
2016-03-27   0.01564
2016-03-28   0.02069
2016-03-29   0.02209
2016-03-30   0.02461
2016-03-31   0.02363
2016-04-01   0.02294
2016-04-02   0.02466
2016-04-03   0.02515
2016-04-04   0.02412
2016-04-05   0.12541
2016-04-06   0.22331
2016-04-07   0.13275
Name: last_seen, dtype: float64

In [27]:
#further analysis of ad_created column
autos["ad_created"].value_counts(normalize=True,ascending=True)

2015-09-09   0.00002
2016-02-01   0.00002
2016-01-14   0.00002
2015-12-05   0.00002
2016-02-17   0.00002
               ...  
2016-03-12   0.03665
2016-04-04   0.03695
2016-03-21   0.03753
2016-03-20   0.03807
2016-04-03   0.03901
Name: ad_created, Length: 74, dtype: float64

From the analysis above, `date_crawled` and `last_seen` dates seem to coincide with each other, indicating that the site is relatively active as the `last_seen` of most ads are quite recent. The `ad_created` date has a larger date range, with a larger proportion of the dates in March and April. This could mean that March and April is the peak period for the car market.

## Exploring the brand column

In [28]:
print(autos["brand"].value_counts(normalize=True)*100,"\n")
print("Number of unique brands: ",len(autos["brand"].unique()))

volkswagen       21.12682
bmw              11.00471
opel             10.75835
mercedes_benz     9.64653
audi              8.65681
ford              6.99015
renault           4.71294
peugeot           2.98415
fiat              2.56427
seat              1.82734
skoda             1.64096
nissan            1.52742
mazda             1.51885
smart             1.41602
citroen           1.40103
toyota            1.27035
hyundai           1.00257
sonstige_autos    0.98115
volvo             0.91474
mini              0.87618
mitsubishi        0.82262
honda             0.78406
kia               0.70694
alfa_romeo        0.66410
porsche           0.61268
suzuki            0.59340
chevrolet         0.56984
chrysler          0.35133
dacia             0.26350
daihatsu          0.25064
jeep              0.22708
subaru            0.21422
land_rover        0.20994
saab              0.16495
jaguar            0.15638
daewoo            0.14996
trabant           0.13925
rover             0.13282
lancia      

In [29]:
top_10_car_brands= autos["brand"].value_counts().head(10).index
car_brand_price={}
car_brand_mileage={}
for brand in top_10_car_brands:
    #create brand_group dataframe
    brand_group = autos.loc[autos["brand"]==brand,:]
    
    #sum of price/mileage column divide by number of rows
    mean_price = brand_group["price"].sum()/brand_group.shape[0]
    mean_mileage = brand_group["odometer_km"].sum()/brand_group.shape[0]
    
    #assign to dictionary
    car_brand_price[brand]=mean_price
    car_brand_mileage[brand]=mean_mileage

#assign into list of tuples to sort
price_list=[]
print("Car brands ranking by avg price:")
for k,v in car_brand_price.items():
    price_list.append((v,k))

price_list.sort(reverse=True)
for x in range(len(price_list)):
    print(price_list[x][1],":",price_list[x][0])
    
print("\n")    

mile_list=[]
print("Car brands ranking by avg mileage:")
for k,v in car_brand_mileage.items():
    mile_list.append((v,k))

mile_list.sort(reverse=True)
for x in range(len(mile_list)):
    print(mile_list[x][1],":",mile_list[x][0])

Car brands ranking by avg price:
audi : 9336.687453600594
mercedes_benz : 8628.450366422385
bmw : 8332.820517811953
volkswagen : 5402.410261610221
seat : 4397.230949589683
ford : 3749.4695065890287
peugeot : 3094.0172290021537
opel : 2975.2419354838707
fiat : 2813.748538011696
renault : 2475.944090909091


Car brands ranking by avg mileage:
bmw : 132572.51313996495
mercedes_benz : 130788.36331334666
opel : 129310.0358422939
audi : 129157.38678544914
volkswagen : 128707.15879132022
renault : 128061.36363636363
peugeot : 127153.62526920316
ford : 124266.01287159056
seat : 121131.30128956624
fiat : 117121.9715956558


In [30]:
#create new dataframe object to analyse correlation of price and mileage
price_series = pd.Series(car_brand_price)
mean_df= pd.DataFrame(price_series, columns=["mean_price"])
mileage_series = pd.Series(car_brand_mileage)
mileage_df= pd.DataFrame(mileage_series, columns=["mean_mileage"])
mean_df=mean_df.join(mileage_df)

#sort by mean_price column
mean_df.sort_values(by="mean_price", ascending=False)


mean_price  mean_mileage
audi          9,336.68745 129,157.38679
mercedes_benz 8,628.45037 130,788.36331
bmw           8,332.82052 132,572.51314
volkswagen    5,402.41026 128,707.15879
seat          4,397.23095 121,131.30129
ford          3,749.46951 124,266.01287
peugeot       3,094.01723 127,153.62527
opel          2,975.24194 129,310.03584
fiat          2,813.74854 117,121.97160
renault       2,475.94409 128,061.36364

In [31]:
#sort by mean_price column
mean_df.sort_values(by="mean_mileage", ascending=False)

mean_price  mean_mileage
bmw           8,332.82052 132,572.51314
mercedes_benz 8,628.45037 130,788.36331
opel          2,975.24194 129,310.03584
audi          9,336.68745 129,157.38679
volkswagen    5,402.41026 128,707.15879
renault       2,475.94409 128,061.36364
peugeot       3,094.01723 127,153.62527
ford          3,749.46951 124,266.01287
seat          4,397.23095 121,131.30129
fiat          2,813.74854 117,121.97160

From the aggregated data, we can see that the mean price of `audi` is the highest amongst the top 10 car brands. In terms of mileage, `bmw` ranks as the highest amongst the top 10 car brands.

We can split the car brand pricing into 3 tiers:

Tier 1 >\\$6000 

Tier 2 <\$6000 and >\$3000 

Tier 3 <\$3000 

Amongst the Tier 1 cars, it seems like there might be an inverse relationship between mean_price and mean_mileage as `audi` with the highest mean_price has the lowest mean_mileage and `bmw` with the lowest mean_price has the highest mean_mileage. 

In Tier 2, there is no correlation between mean_price and mean_mileage as `volkswagen` being the highest in this price range has the highest mean_mileage as well. Additionally, `peugeot` which has the lowest mean_price in this tier, has the second highest mean_mileage. This shows there is no suggestion of correlation between the 2 metrics.

In tier 3, there is also no correlation as `opel` with the highest mean_price, also has the highest mean_mileage. 

Across the board, as we look at the table sorted by mean_mileage, there is no obvious relation between mean_price and mean_mileage. This suggests that car buyers believe in the intrinsic value of a car brand and are less affected by the mileage on the car. 


## Additional metrics
Diving further, let us look at the most common brand/model combinations.

In [32]:
brand_model=autos["brand"]+"-"+autos["model"]
brand_model_list=brand_model.value_counts().index
avg_price_list={}
for brand_mod in brand_model_list:
    brand_model_df = autos.loc[autos["brand"]+"-"+autos["model"]==brand_mod,:]
    
    avg_price = brand_model_df["price"].sum()/brand_model_df.shape[0]
    avg_price_list[brand_mod]=avg_price

#create dataframe for brand_model_price    
brand_model_price = pd.Series(avg_price_list)
brand_model_price_df = pd.DataFrame(brand_model_price, columns=["mean_price"])

#create dataframe for brand_model_count    
brand_model_count = pd.Series(brand_model.value_counts())
brand_model_count_df = pd.DataFrame(brand_model_count, columns=["count"])

#join the 2 dataframes
brand_model_price_df = brand_model_price_df.join(brand_model_count_df)

brand_model_price_df.sort_values(by="count", ascending=False)

mean_price  count
volkswagen-golf    5,081.57998   3707
bmw-3er            6,001.79273   2615
volkswagen-polo    2,593.38347   1609
opel-corsa         1,876.50628   1592
volkswagen-passat  4,948.30096   1349
...                        ...    ...
ford-b_max        10,600.00000      1
bmw-i3               250.00000      1
rover-discovery    2,800.00000      1
rover-rangerover  23,500.00000      1
audi-200           2,250.00000      1

[290 rows x 2 columns]

The most common brand/model combination listed would be a volksawagen-gold with an average selling price of \\$5081.58

Now let us split the data on the different car models and compare prices for damaged cars with non-damaged cars.

In [33]:
#convert german to english
autos.loc[autos["unrepaired_damage"]=="nein","unrepaired_damage"]="no"
autos.loc[autos["unrepaired_damage"]=="ja","unrepaired_damage"]="yes"

autos["unrepaired_damage"].value_counts()

no     33834
yes     4540
Name: unrepaired_damage, dtype: int64

In [39]:
brand_model=autos["brand"]+"-"+autos["model"]
#list of unique brand_models
brand_model_list=brand_model.value_counts().index

avg_price_dict={}
avg_price_dmg_dict={}

#for-loop to calculate the avg price for dmged/non-dmged car of the same brand_model
for brand_mod in brand_model_list:
    brand_model_df = autos.loc[(autos["brand"]+"-"+autos["model"]==brand_mod)&(autos["unrepaired_damage"]=="no"),:]
    brand_model_dmged_df = autos.loc[(autos["brand"]+"-"+autos["model"]==brand_mod)&(autos["unrepaired_damage"]=="yes"),:]
                                                                               
    avg_price = brand_model_df["price"].sum()/brand_model_df.shape[0]
    avg_price_dmged = brand_model_dmged_df["price"].mean()
    
    avg_price_dict[brand_mod]=avg_price
    avg_price_dmg_dict[brand_mod]=avg_price_dmged

#create dataframe for brand_model_price    
brand_model_price = pd.Series(avg_price_dict)
brand_model_price_df = pd.DataFrame(brand_model_price, columns=["mean_price_no_dmg"])

#create dataframe for brand_model_price_dmg
brand_model_price_dmg = pd.Series(avg_price_dmg_dict)
brand_model_price_dmg_df = pd.DataFrame(brand_model_price_dmg, columns=["mean_price_dmged"])
                                                                                     
#join the 2 dataframes
brand_model_price_df = brand_model_price_df.join(brand_model_price_dmg_df)

#remove brand/model with no mean_price_dmged data
brand_model_price_df = brand_model_price_df.dropna(axis=0)

#join price_diff column with brand_model_price_df
price_diff = pd.DataFrame(brand_model_price_df["mean_price_no_dmg"]-brand_model_price_df["mean_price_dmged"], columns=["mean_price_diff"])
brand_model_price_df=brand_model_price_df.join(price_diff)

brand_model_price_df.sort_values(by="mean_price_diff", ascending=False)

<ipython-input-39-11bcc33e13e8>:13: RuntimeWarning: invalid value encountered in longlong_scalars
  avg_price = brand_model_df["price"].sum()/brand_model_df.shape[0]


mean_price_no_dmg  mean_price_dmged  mean_price_diff
porsche-911                  74,240.92248      20,833.33333     53,407.58915
land_rover-range_rover       25,411.42857       1,800.00000     23,611.42857
land_rover-defender          33,009.12500      12,245.00000     20,764.12500
bmw-m_reihe                  28,971.97561       8,900.00000     20,071.97561
mercedes_benz-g_klasse       25,956.73684       6,375.00000     19,581.73684
...                                   ...               ...              ...
mercedes_benz-cl             13,229.33333      14,425.00000     -1,195.66667
volkswagen-andere             5,564.45714       7,116.66667     -1,552.20952
lancia-andere                 4,913.00000       8,750.00000     -3,837.00000
mercedes_benz-viano          20,365.25455      24,333.33333     -3,968.07879
volkswagen-touareg           16,169.65060      20,933.33333     -4,763.68273

[238 rows x 3 columns]

We can see that the range of `mean_price_diff` is quite large, ranging from \\$53,407 to -\$4,763. In general, damaged cars are cheaper in comparison with their non-damaged counterparts. However this is dependent on the magnitude of damage and the cost of repair. Hence, we are not able to draw a valid conclusion with this metric.

# Conclusion
We explored and cleaned the data from eBay Kleinanzeigen by doing the following:
1. Renamed camelcase to snakecase
2. Changed numeric data from object dtype to int/float dtype for `price` and `odometer_km` columns
3. Removed illogical data from `registration_year` column
4. Removed single valued columns (`seller`, `offer_type` and `nr_of_pictures`)
5. Removed time data from `date_crawled`, `last_seen`, and `ad_created` columns

We also developed several insights:
1. The most popular brand on the site is `volkswagen` taking up 21.1% of cars listed on the website
2. There is no obvious relation between mean_price and mean_mileage. 
3. Car buyers believe in the intrinsic value of a car brand and are less affected by the mileage on the car.
4. The top 3 brand/model listed are `volkswagen-golf`, `bmw-3er` and `volkswagen-polo`.
5. Damaged cars are generally priced cheaper than non-damaged cars with the exception of several brand/model. We believe this is greatly influenced by the extent of damage and are unable to analyse any further based on the given data.